File: cmd/dilithium-gpu-miner/miner.go
package main

import (
	"context"
	"crypto/sha256"
	"encoding"
	"encoding/json"
	"fmt"
	"runtime"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

// Miner coordinates CPU or GPU workers to mine dilithium blocks.
type Miner struct {
	client  *NodeClient
	address string
	threads int

	// GPU settings
	useGPU    bool
	gpuDevice int
	batchSize uint64

	// Stats
	blocksMined atomic.Int64
	totalHashes atomic.Int64
	earnings    atomic.Int64
	startTime   time.Time

	// Control
	stopCh chan struct{}
	wg     sync.WaitGroup
}

// NewMiner creates a new mining coordinator.
func NewMiner(nodeURL, address string, threads int) *Miner {
	if threads <= 0 {
		threads = runtime.NumCPU()
	}
	return &Miner{
		client:  NewNodeClient(nodeURL),
		address: address,
		threads: threads,
		stopCh:  make(chan struct{}),
	}
}

// Start begins the mining loop.
func (m *Miner) Start() error {
	if err := m.client.CheckConnection(); err != nil {
		return err
	}

	// Initialize GPU if needed
	if m.useGPU {
		if !GPUMiningAvailable {
			return fmt.Errorf("GPU mining not available - rebuild with CUDA support")
		}
		if err := GPUInit(m.gpuDevice); err != nil {
			return fmt.Errorf("failed to initialize GPU: %w", err)
		}
	}

	m.startTime = time.Now()

	// Start stats reporter
	m.wg.Add(1)
	go m.statsLoop()

	// Start main mining loop
	m.wg.Add(1)
	go m.miningLoop()

	return nil
}

// Stop halts all mining activity.
func (m *Miner) Stop() {
	close(m.stopCh)
	m.wg.Wait()
	if m.useGPU {
		GPUCleanup()
	}
}

// waitForSync blocks until the node is synced with the real network.
// Requires: (1) at least one peer connected, (2) height > 0, (3) height stable for 10 seconds.
func (m *Miner) waitForSync() {
	fmt.Println("[*] Waiting for node to connect to peers and sync...")

	// Phase 1: Wait for at least one peer
	for {
		select {
		case <-m.stopCh:
			return
		default:
		}

		peers, err := m.client.GetPeerCount()
		if err == nil && peers > 0 {
			fmt.Printf("[+] Connected to %d peer(s)\n", peers)
			break
		}
		fmt.Println("[~] No peers yet, waiting...")
		m.sleep(3 * time.Second)
	}

	// Phase 2: Wait for height > 0 (chain data received from peers)
	for {
		select {
		case <-m.stopCh:
			return
		default:
		}

		height, err := m.client.GetCurrentHeight()
		if err == nil && height > 0 {
			fmt.Printf("[~] Chain height: %d, waiting for sync to complete...\n", height)
			break
		}
		m.sleep(2 * time.Second)
	}

	// Phase 3: Wait for height to stabilize (not advancing for 5 checks = 10 seconds)
	var lastHeight int64
	stableCount := 0

	for {
		select {
		case <-m.stopCh:
			return
		default:
		}

		height, err := m.client.GetCurrentHeight()
		if err != nil {
			m.sleep(2 * time.Second)
			continue
		}

		if height == lastHeight {
			stableCount++
		} else {
			if lastHeight > 0 {
				fmt.Printf("[~] Syncing... height: %d (+%d)\n", height, height-lastHeight)
			}
			stableCount = 0
			lastHeight = height
		}

		// Height stable for 5 checks (10 seconds) = likely synced
		if stableCount >= 5 {
			fmt.Printf("[+] Node synced at height %d\n", height)
			return
		}

		m.sleep(2 * time.Second)
	}
}

// miningLoop continuously fetches work and mines blocks.
// Includes pending mempool transactions alongside the coinbase reward.
func (m *Miner) miningLoop() {
	defer m.wg.Done()

	// Wait for node to finish syncing before mining
	m.waitForSync()

	for {
		select {
		case <-m.stopCh:
			return
		default:
		}

		// Fetch work template
		template, err := m.client.GetWork()
		if err != nil {
			fmt.Printf("[!] Error getting work: %v\n", err)
			m.sleep(5 * time.Second)
			continue
		}

		// Fetch pending transactions from mempool
		pendingTxs := m.client.GetPendingTransactions()
		if len(pendingTxs) > 0 {
			fmt.Printf("[*] Including %d mempool transaction(s) in block\n", len(pendingTxs))
		}

		// Build block with coinbase + pending transactions
		block := m.constructBlock(template, pendingTxs)

		// Prepare mining data: compute prefix, suffix, and midstate
		prefix, suffix := m.buildHashInput(block)

		// Determine difficulty
		diffBits := template.DifficultyBits
		if diffBits <= 0 {
			diffBits = template.Difficulty * 4
		}

		if m.useGPU {
			fmt.Printf("[*] Mining block #%d | difficulty: %d bits (%d hex) | GPU device %d\n",
				template.Index, template.DifficultyBits, template.Difficulty, m.gpuDevice)
		} else {
			if template.DifficultyBits > 0 {
				fmt.Printf("[*] Mining block #%d | difficulty: %d bits (%d hex) | %d threads\n",
					template.Index, template.DifficultyBits, template.Difficulty, m.threads)
			} else {
				fmt.Printf("[*] Mining block #%d | difficulty: %d hex digits | %d threads\n",
					template.Index, template.Difficulty, m.threads)
			}
		}

		// Mine!
		result, found := m.mineWithWorkers(prefix, suffix, diffBits, template.Height)
		if !found {
			continue
		}

		// Set block fields from result
		block.Nonce = result.Nonce
		block.Hash = hashToHex(result.Hash)

		// Verify hash matches what the node expects (sanity check)
		expectedHash := m.calculateBlockHashGo(block)
		if block.Hash != expectedHash {
			fmt.Printf("[!] HASH MISMATCH: computed=%s expected=%s\n", block.Hash, expectedHash)
			fmt.Printf("[!] This is a bug — skipping submission\n")
			continue
		}

		// Submit block
		if err := m.client.SubmitBlock(block); err != nil {
			fmt.Printf("[!] Block rejected: %v\n", err)
			continue
		}

		// Block accepted!
		m.blocksMined.Add(1)
		m.earnings.Add(template.Reward)
		fmt.Printf("[+] BLOCK #%d MINED AND ACCEPTED! Hash: %s... | Reward: %s DLT | Total: %d blocks, %s DLT\n",
			block.Index, block.Hash[:16],
			FormatDLT(template.Reward),
			m.blocksMined.Load(), FormatDLT(m.earnings.Load()))
	}
}

// constructBlock builds a block with coinbase + pending transactions, ready for mining.
func (m *Miner) constructBlock(template *BlockTemplate, pendingTxs []*Transaction) *Block {
	coinbase := &Transaction{
		From:      "SYSTEM",
		To:        m.address,
		Amount:    template.Reward,
		Timestamp: time.Now().Unix(),
		Signature: fmt.Sprintf("coinbase-%d-%d", template.Index, time.Now().UnixNano()),
	}

	txs := make([]*Transaction, 0, 1+len(pendingTxs))
	txs = append(txs, coinbase)
	txs = append(txs, pendingTxs...)

	return &Block{
		Index:          template.Index,
		Timestamp:      time.Now().Unix(),
		Transactions:   txs,
		PreviousHash:   template.PreviousHash,
		Difficulty:     template.Difficulty,
		DifficultyBits: template.DifficultyBits,
	}
}

// buildHashInput computes the fixed prefix and suffix for hash computation.
// Block hash = SHA256(Index + Timestamp + txJSON + PreviousHash + Nonce + Difficulty)
// prefix = everything before Nonce, suffix = everything after Nonce.
func (m *Miner) buildHashInput(block *Block) (prefix, suffix []byte) {
	txJSON, _ := json.Marshal(block.Transactions)

	prefixStr := strconv.FormatInt(block.Index, 10) +
		strconv.FormatInt(block.Timestamp, 10) +
		string(txJSON) +
		block.PreviousHash

	suffixStr := strconv.Itoa(block.Difficulty)

	return []byte(prefixStr), []byte(suffixStr)
}

// calculateBlockHashGo computes the block hash using the same method as the node.
// Used only for verification — not in the hot loop.
func (m *Miner) calculateBlockHashGo(block *Block) string {
	txJSON, _ := json.Marshal(block.Transactions)
	blockData := strconv.FormatInt(block.Index, 10) +
		strconv.FormatInt(block.Timestamp, 10) +
		string(txJSON) +
		block.PreviousHash +
		strconv.FormatInt(block.Nonce, 10) +
		strconv.Itoa(block.Difficulty)

	state := InitSHA256()
	data := []byte(blockData)
	fullBlocks := (len(data) / 64) * 64
	if fullBlocks > 0 {
		state.ProcessBlocks(data[:fullBlocks])
	}
	hash := MineHash(state.H, data[fullBlocks:], state.Len)
	return hashToHex(hash)
}

// mineWithWorkers launches CPU or GPU workers to search for a valid nonce.
func (m *Miner) mineWithWorkers(prefix, suffix []byte, diffBits int, templateHeight int64) (MiningResult, bool) {
	// Compute SHA-256 midstate using stdlib (hardware-accelerated)
	h := sha256.New()
	fullBlockLen := (len(prefix) / 64) * 64
	if fullBlockLen > 0 {
		h.Write(prefix[:fullBlockLen])
	}
	midstate, _ := h.(encoding.BinaryMarshaler).MarshalBinary()
	prefixTail := prefix[fullBlockLen:]

	fmt.Printf("[*] Prefix: %d bytes | Midstate: %d blocks skipped | Tail: %d bytes\n",
		len(prefix), fullBlockLen/64, len(prefixTail))

	// Create cancellable context
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	resultCh := make(chan MiningResult, 1)

	if m.useGPU {
		// Launch single GPU worker
		worker := &GPUWorker{
			ID:        0,
			DeviceID:  m.gpuDevice,
			BatchSize: m.batchSize,
		}
		go func() {
			worker.Mine(ctx, midstate, prefixTail, suffix, 0, 1, diffBits, resultCh)
		}()

		// Monitor for results or stale work
		ticker := time.NewTicker(3 * time.Second)
		defer ticker.Stop()

		hashStart := m.totalHashes.Load()
		startTime := time.Now()

		for {
			select {
			case result := <-resultCh:
				cancel()
				m.totalHashes.Add(worker.HashCount.Load())

				elapsed := time.Since(startTime).Seconds()
				hashes := m.totalHashes.Load() - hashStart
				if elapsed > 0 {
					fmt.Printf("[+] Found hash after %d hashes (%.2f MH/s)\n",
						hashes, float64(hashes)/elapsed/1e6)
				}
				return result, true

			case <-ticker.C:
				currentHeight, err := m.client.GetCurrentHeight()
				if err == nil && currentHeight > templateHeight {
					fmt.Printf("[~] Chain advanced to %d, restarting with new work...\n", currentHeight)
					cancel()
					m.totalHashes.Add(worker.HashCount.Load())
					return MiningResult{}, false
				}

				hashes := worker.HashCount.Load()
				elapsed := time.Since(startTime).Seconds()
				if elapsed > 0 {
					rate := float64(hashes) / elapsed
					if rate > 1e6 {
						fmt.Printf("[~] Hashrate: %.2f MH/s | Hashes: %d\n", rate/1e6, hashes)
					} else {
						fmt.Printf("[~] Hashrate: %.0f KH/s | Hashes: %d\n", rate/1e3, hashes)
					}
				}

			case <-m.stopCh:
				cancel()
				m.totalHashes.Add(worker.HashCount.Load())
				return MiningResult{}, false
			}
		}
	}

	// CPU mining
	var workerWg sync.WaitGroup
	workers := make([]*CPUWorker, m.threads)

	for i := 0; i < m.threads; i++ {
		workers[i] = &CPUWorker{ID: i}
		workerWg.Add(1)
		go func(w *CPUWorker, startNonce int64) {
			defer workerWg.Done()
			w.Mine(ctx, midstate, prefixTail, suffix,
				startNonce, int64(m.threads), diffBits, resultCh)
		}(workers[i], int64(i))
	}

	// Monitor for results, stale work, or shutdown
	ticker := time.NewTicker(3 * time.Second)
	defer ticker.Stop()

	hashStart := m.totalHashes.Load()
	startTime := time.Now()

	for {
		select {
		case result := <-resultCh:
			cancel()
			workerWg.Wait()

			// Accumulate hash counts
			for _, w := range workers {
				m.totalHashes.Add(w.HashCount.Load())
			}

			elapsed := time.Since(startTime).Seconds()
			hashes := m.totalHashes.Load() - hashStart
			if elapsed > 0 {
				fmt.Printf("[+] Found hash after %d hashes (%.2f MH/s)\n",
					hashes, float64(hashes)/elapsed/1e6)
			}
			return result, true

		case <-ticker.C:
			// Check if chain has advanced (stale work detection)
			currentHeight, err := m.client.GetCurrentHeight()
			if err == nil && currentHeight > templateHeight {
				fmt.Printf("[~] Chain advanced to %d, restarting with new work...\n", currentHeight)
				cancel()
				workerWg.Wait()
				for _, w := range workers {
					m.totalHashes.Add(w.HashCount.Load())
				}
				return MiningResult{}, false
			}

			// Print live hashrate
			var totalWorkerHashes int64
			for _, w := range workers {
				totalWorkerHashes += w.HashCount.Load()
			}
			elapsed := time.Since(startTime).Seconds()
			if elapsed > 0 {
				rate := float64(totalWorkerHashes) / elapsed
				if rate > 1e6 {
					fmt.Printf("[~] Hashrate: %.2f MH/s | Hashes: %d\n", rate/1e6, totalWorkerHashes)
				} else {
					fmt.Printf("[~] Hashrate: %.0f KH/s | Hashes: %d\n", rate/1e3, totalWorkerHashes)
				}
			}

		case <-m.stopCh:
			cancel()
			workerWg.Wait()
			for _, w := range workers {
				m.totalHashes.Add(w.HashCount.Load())
			}
			return MiningResult{}, false
		}
	}
}

// statsLoop prints periodic statistics.
func (m *Miner) statsLoop() {
	defer m.wg.Done()

	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			elapsed := time.Since(m.startTime).Seconds()
			totalH := m.totalHashes.Load()
			blocks := m.blocksMined.Load()
			earn := m.earnings.Load()

			fmt.Printf("[i] Session: %.0fs | Total hashes: %d | Blocks: %d | Earnings: %s DLT\n",
				elapsed, totalH, blocks, FormatDLT(earn))

		case <-m.stopCh:
			return
		}
	}
}

// sleep waits for duration or until stopped.
func (m *Miner) sleep(d time.Duration) {
	select {
	case <-time.After(d):
	case <-m.stopCh:
	}
}

File: cmd/dilithium-gpu-miner/src/miner.rs
// Mining coordinator

use crate::backend::{self, GpuBackend};
use crate::network::{Block, BlockTemplate, NodeClient, Transaction};
use crate::sha256::sha256_midstate;
use crate::webui::MinerStats;
use crate::worker::GpuWorker;
use anyhow::{Context, Result};
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::sync::Arc;
use std::time::{Duration, Instant};
use tokio::time::sleep;

#[derive(Clone)]
pub struct Miner {
    node_url: String,
    wallet_address: String,
    device_id: i32,
    batch_size: u64,

    // Stats
    blocks_mined: Arc<AtomicU64>,
    total_hashes: Arc<AtomicU64>,
    total_earnings: Arc<AtomicU64>,
    current_hashrate: Arc<AtomicU64>, // Real-time hashrate in H/s
    stop_flag: Arc<AtomicBool>,
}

impl Miner {
    pub fn new(node_url: String, wallet_address: String, device_id: i32, batch_size: u64) -> Self {
        Self {
            node_url,
            wallet_address,
            device_id,
            batch_size,
            blocks_mined: Arc::new(AtomicU64::new(0)),
            total_hashes: Arc::new(AtomicU64::new(0)),
            total_earnings: Arc::new(AtomicU64::new(0)),
            current_hashrate: Arc::new(AtomicU64::new(0)),
            stop_flag: Arc::new(AtomicBool::new(false)),
        }
    }

    pub fn stop(&self) {
        self.stop_flag.store(true, Ordering::Relaxed);
    }

    pub fn get_stats(&self, start_time: Instant) -> MinerStats {
        MinerStats {
            blocks_mined: self.blocks_mined.clone(),
            total_hashes: self.total_hashes.clone(),
            total_earnings: self.total_earnings.clone(),
            current_hashrate: self.current_hashrate.clone(),
            start_time,
            wallet_address: self.wallet_address.clone(),
            node_url: self.node_url.clone(),
            device_id: self.device_id,
            batch_size: self.batch_size,
        }
    }

    pub async fn run(&self) -> Result<()> {
        let client = NodeClient::new(self.node_url.clone());

        // Check connection
        client
            .check_connection()
            .await
            .context("Failed to connect to node")?;

        log::info!("[*] Connected to node at {}", self.node_url);
        log::info!("[*] Mining to wallet: {}", self.wallet_address);

        // Initialize GPU backend
        let gpu_backend = backend::create_backend(self.device_id)
            .context("Failed to initialize GPU backend")?;
        log::info!(
            "[GPU {}] {} backend initialized with {} compute units, batch size: {}",
            self.device_id,
            gpu_backend.name(),
            gpu_backend.compute_unit_count(),
            self.batch_size
        );

        // Wrap in Arc for sharing with spawn_blocking
        let gpu_backend: Arc<dyn GpuBackend> = Arc::from(gpu_backend);

        // Wait for sync
        self.wait_for_sync(&client).await?;

        // Start stats reporter
        let stats_handle = {
            let blocks = self.blocks_mined.clone();
            let hashes = self.total_hashes.clone();
            let earnings = self.total_earnings.clone();
            let stop = self.stop_flag.clone();

            tokio::spawn(async move {
                let start_time = Instant::now();

                while !stop.load(Ordering::Relaxed) {
                    sleep(Duration::from_secs(30)).await;

                    let elapsed = start_time.elapsed().as_secs();
                    let total_h = hashes.load(Ordering::Relaxed);
                    let blocks_count = blocks.load(Ordering::Relaxed);
                    let earn = earnings.load(Ordering::Relaxed);

                    let avg_hashrate = if elapsed > 0 {
                        total_h as f64 / elapsed as f64 / 1e6
                    } else {
                        0.0
                    };

                    log::info!(
                        "[i] Session: {}s | Hashes: {} | Avg: {:.2} MH/s | Blocks: {} | Earnings: {} DLT",
                        elapsed,
                        total_h,
                        avg_hashrate,
                        blocks_count,
                        earn as f64 / 1e8
                    );
                }
            })
        };

        // Main mining loop
        loop {
            if self.stop_flag.load(Ordering::Relaxed) {
                break;
            }

            // Get work
            let template = match client.get_work().await {
                Ok(t) => t,
                Err(e) => {
                    log::error!("[!] Error getting work: {}", e);
                    sleep(Duration::from_secs(5)).await;
                    continue;
                }
            };

            // Get pending transactions
            let pending_txs = client.get_pending_transactions().await.unwrap_or_default();

            if !pending_txs.is_empty() {
                log::info!("[*] Including {} pending transaction(s)", pending_txs.len());
            }

            // Build block
            let block = self.construct_block(&template, pending_txs);

            log::info!(
                "[*] Mining block #{} | difficulty: {} bits ({} hex) | {} backend",
                template.index,
                template.difficulty_bits,
                template.difficulty,
                gpu_backend.name()
            );

            // Mine block
            match self.mine_block(&block, &template, &gpu_backend).await {
                Ok(Some((nonce, hash, solved_block))) => {
                    // Verify block is still valid (chain might have advanced)
                    let current_height = client.get_current_height().await.unwrap_or(template.index);
                    if current_height != template.index {
                        log::info!("[~] Block stale (height {} -> {}), restarting...", template.index, current_height);
                        continue;
                    }

                    // Set block fields — use solved_block which has the correct timestamp
                    let mut mined_block = solved_block;
                    mined_block.nonce = nonce as i64;
                    mined_block.hash = hex::encode(hash);

                    log::info!("[*] Submitting block #{} (height {}, prev: {}...)",
                        mined_block.index, template.height, &template.previous_hash[..16]);

                    // Submit block
                    match client.submit_block(&mined_block).await {
                        Ok(_) => {
                            self.blocks_mined.fetch_add(1, Ordering::Relaxed);
                            self.total_earnings
                                .fetch_add(template.reward as u64, Ordering::Relaxed);

                            log::info!(
                                "[+] BLOCK #{} MINED! Hash: {}... | Reward: {} DLT | Total: {} blocks, {} DLT",
                                mined_block.index,
                                &mined_block.hash[..16],
                                template.reward as f64 / 1e8,
                                self.blocks_mined.load(Ordering::Relaxed),
                                self.total_earnings.load(Ordering::Relaxed) as f64 / 1e8
                            );
                        }
                        Err(e) => {
                            log::error!("[!] Block rejected: {}", e);
                            // Block rejected, restart with fresh template
                            continue;
                        }
                    }
                }
                Ok(None) => {
                    // Stale work, restart
                    continue;
                }
                Err(e) => {
                    log::error!("[!] Mining error: {}", e);
                    sleep(Duration::from_secs(1)).await;
                }
            }
        }

        stats_handle.abort();
        Ok(())
    }

    async fn wait_for_sync(&self, client: &NodeClient) -> Result<()> {
        log::info!("[*] Checking node sync status...");

        // Just wait for height > 0, don't require stability
        loop {
            if self.stop_flag.load(Ordering::Relaxed) {
                return Ok(());
            }

            let height = match client.get_current_height().await {
                Ok(h) => h,
                Err(e) => {
                    log::error!("[!] Error getting height: {}", e);
                    0
                }
            };

            if height > 0 {
                log::info!("[+] Node ready at height {}", height);
                return Ok(());
            }

            log::info!("[~] Waiting for blockchain data...");
            sleep(Duration::from_secs(2)).await;
        }
    }

    fn construct_block(&self, template: &BlockTemplate, pending_txs: Vec<Transaction>) -> Block {
        let coinbase = Transaction {
            from: "SYSTEM".to_string(),
            to: self.wallet_address.clone(),
            amount: template.reward,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs() as i64,
            signature: format!(
                "coinbase-{}-{}",
                template.index,
                std::time::SystemTime::now()
                    .duration_since(std::time::UNIX_EPOCH)
                    .unwrap()
                    .as_nanos()
            ),
        };

        let mut txs = vec![coinbase];
        txs.extend(pending_txs);

        Block {
            index: template.index,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs() as i64,
            transactions: txs,
            previous_hash: template.previous_hash.clone(),
            hash: String::new(),
            nonce: 0,
            difficulty: template.difficulty,
            difficulty_bits: Some(template.difficulty_bits),
        }
    }

    async fn mine_block(
        &self,
        block: &Block,
        template: &BlockTemplate,
        gpu_backend: &Arc<dyn GpuBackend>,
    ) -> Result<Option<(u64, [u8; 32], Block)>> {
        let start_time = Instant::now();
        let node_client = NodeClient::new(self.node_url.clone());
        let cumulative_hashes = Arc::new(AtomicU64::new(0));
        let base_total_hashes = self.total_hashes.load(Ordering::Relaxed);

        // Mutable block for timestamp rotation
        let mut rotated_block = block.clone();

        loop {
            // Build hash input with current timestamp
            let (prefix, suffix) = self.build_hash_input(&rotated_block);

            // Compute midstate
            let midstate = sha256_midstate(&prefix);
            let full_blocks_len = (prefix.len() / 64) * 64;
            let tail = &prefix[full_blocks_len..];

            // Compute max nonce to stay on single-block SHA-256 fast path
            let max_nonce_digits = 55_i32 - tail.len() as i32 - suffix.len() as i32;
            let max_nonce = if max_nonce_digits >= 1 {
                10u64.pow(max_nonce_digits as u32) - 1
            } else {
                u64::MAX
            };

            // Create worker for this rotation
            let hash_count = Arc::new(AtomicU64::new(0));
            let worker_stop = Arc::new(AtomicBool::new(false));

            let worker = GpuWorker::new(
                self.batch_size,
                max_nonce,
                hash_count.clone(),
                worker_stop.clone(),
            );

            let midstate_copy = midstate;
            let tail_copy = tail.to_vec();
            let suffix_copy = suffix.clone();
            let prefix_len = prefix.len() as u64;
            let diff_bits = template.difficulty_bits;
            let backend_clone = gpu_backend.clone();

            let mut mining_handle = tokio::task::spawn_blocking(move || {
                worker.mine(
                    backend_clone.as_ref(),
                    &midstate_copy,
                    &tail_copy,
                    prefix_len,
                    &suffix_copy,
                    diff_bits,
                )
            });

            // Monitor for stale work or completion
            loop {
                tokio::select! {
                    result = &mut mining_handle => {
                        let hashes = hash_count.load(Ordering::Relaxed);
                        cumulative_hashes.fetch_add(hashes, Ordering::Relaxed);

                        match result {
                            Ok(Ok(Some(solution))) => {
                                let total = cumulative_hashes.load(Ordering::Relaxed);
                                self.total_hashes.store(base_total_hashes + total, Ordering::Relaxed);
                                let elapsed = start_time.elapsed().as_secs_f64();
                                let hashrate = if elapsed > 0.0 { total as f64 / elapsed / 1e6 } else { 0.0 };
                                log::info!("[+] Found hash after {} hashes ({:.2} MH/s)", total, hashrate);
                                return Ok(Some((solution.nonce, solution.hash, rotated_block.clone())));
                            }
                            Ok(Ok(None)) => {
                                if worker_stop.load(Ordering::Relaxed) {
                                    // Stale work - return to main loop
                                    let total = cumulative_hashes.load(Ordering::Relaxed);
                                    self.total_hashes.store(base_total_hashes + total, Ordering::Relaxed);
                                    return Ok(None);
                                }
                                // Nonce space exhausted - rotate timestamp and retry
                                // (no network calls needed!)
                                rotated_block.timestamp = std::time::SystemTime::now()
                                    .duration_since(std::time::UNIX_EPOCH)
                                    .unwrap()
                                    .as_secs() as i64;
                                break; // break inner loop, restart with new prefix
                            }
                            Ok(Err(e)) => return Err(e),
                            Err(e) => anyhow::bail!("Worker panic: {}", e),
                        }
                    }

                    _ = sleep(Duration::from_secs(3)) => {
                        // Check if chain advanced
                        if let Ok(current_height) = node_client.get_current_height().await {
                            if current_height > template.index {
                                log::info!("[~] Chain advanced to {}, restarting...", current_height);
                                worker_stop.store(true, Ordering::Relaxed);
                                // Wait for worker to finish
                                let _ = mining_handle.await;
                                let total = cumulative_hashes.load(Ordering::Relaxed)
                                    + hash_count.load(Ordering::Relaxed);
                                self.total_hashes.store(base_total_hashes + total, Ordering::Relaxed);
                                return Ok(None);
                            }
                        }

                        // Print progress (cumulative across rotations)
                        let total = cumulative_hashes.load(Ordering::Relaxed)
                            + hash_count.load(Ordering::Relaxed);
                        // Keep session stats current
                        self.total_hashes.store(base_total_hashes + total, Ordering::Relaxed);
                        let elapsed = start_time.elapsed().as_secs_f64();
                        if elapsed > 0.0 {
                            let hashrate = total as f64 / elapsed;
                            self.current_hashrate.store(hashrate as u64, Ordering::Relaxed);
                            log::info!("[~] Hashrate: {:.2} MH/s | Hashes: {}", hashrate / 1e6, total);
                        }
                    }
                }
            }
        }
    }

    fn build_hash_input(&self, block: &Block) -> (Vec<u8>, Vec<u8>) {
        let tx_json = serde_json::to_string(&block.transactions).unwrap();

        let prefix = format!(
            "{}{}{}{}",
            block.index, block.timestamp, tx_json, block.previous_hash
        );

        let suffix = format!("{}", block.difficulty);

        (prefix.into_bytes(), suffix.into_bytes())
    }
}
